Highlight particular words in a sentence

zfletch opened the pull request:
Fixes https://github.com/perseids-tools/treebank-react/issues/25

Add the `highlight` prop to `<Sentence>`. This prop is an array (that's converted to a Set internally) and is passed to all children. It represents words that should be emphasized in some way.

Update the `<Text>`, `<Xml>`, and `<Graph>` components to change the background color of all highlighted words.

In the screenshot below, the first 11 words (including punctuation) are highlighted,


---

<img width="1086" alt="screenshot" src="https://user-images.githubusercontent.com/3039310/111680902-12fb5600-87f9-11eb-9b51-330e3991081f.png">


---

What this means for the user of the library is that they can highlight words by adding `highlight={["array", "of", "ids"]}` to the `<Sentence>` component. To test it locally, [set up](https://github.com/perseids-tools/treebank-react#development) the local development environment and change line 78 of [src/demo/App/App.js](https://github.com/perseids-tools/treebank-react/blob/master/src/demo/App/App.js#L78) to:

```
<Sentence id="1" highlight={["1", "10", "11"]}>
```

zfletch committed:
highlight particular words in a sentence

Add the `highlight` prop to `<Sentence>`. This prop is an array
(that's converted to a Set internally) and is passed to all
children. It represents words that should be emphasized in some way.

Update the `<Text>`, `<Xml>`, and `<Graph>` components to change the
background color of all highlighted words.

zfletch commented on src/lib/components/Treebank/Sentence/Sentence.js:
```diff
@@ -15,8 +17,16 @@ const findWord = (wordId, sentence) => (
 );
 
 const WrappedSentence = ({
```
This library is built using something called React Contexts. A Context is a way of providing values to child components (and grandchild components, great-grandchild components, etc.) without explicitly passing them in. It's sort of like having a variable be available to all functions within its scope:

```javascript
// This is like a Context Provider
let someVariable = 5;

function () {
  function () {
    function () {
       // This is like a Context Consumer
       // It has access to the `someVariable` variable without it being passed in as an argument
       console.log(someVariable);
    }
  }
}
```

---

In the same way, the `<Sentence>` component is providing variables within its scope. This PR is adding the `highlighted` variable to the Context:

```jsx
<Sentence highlighted={['1', '2']}>
  <div>
    {/* This component has access to `highlighted` even though it is not passed in directly */}
    <Text />
  </div>
</Sentence>
```

---

The reason that this library is set up this way, is it allows the user to structure their application however they want while still having access to all of the components of this library. For example, they could put the `<Text>` below the `<Graph>`, have two separate `<Graph>`s, whatever.

zfletch commented on src/lib/components/Treebank/Sentence/Sentence.js:
```diff
@@ -28,6 +38,8 @@ const WrappedSentence = ({
     [activeId, setActiveId] = useState(null);
   }
 
+  const highlighted = new Set(highlightedIterable);
```
The `highlighted` being passed in can be either an array or a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set). But we convert it to a set to make lookup more efficient (see comment on `Text.js`).

zfletch commented on src/lib/components/Treebank/Sentence/Sentence.js:
```diff
@@ -91,13 +105,15 @@ Sentence.propTypes = {
   callback: func,
   active: string,
   setActive: func,
+  highlighted: oneOfType([arrayOf(string), instanceOf(Set)]),
```
Prop Types are a kind of type checking used by React. Ideally, we'd want to say it's either an array of strings or a Set of strings, but there is no `setOf` type. `instanceOf(Set)` is, I think, the next best thing.

zfletch commented on src/lib/components/Treebank/Text/Text.js:
```diff
@@ -7,9 +7,15 @@ import Text from '../../Text';
   <SentenceContext.Consumer>
     {({
-      sentence, active, toggleActive, config,
+      sentence, active, toggleActive, highlighted, config,
```
It's a little confusing that there's a `src/lib/components/Treebank/Text/Text.js` file and a `src/lib/components/Text/Text.js` file.

The reason for this goes back to React Contexts (mentioned above). `src/lib/components/Text/Text.js` is just a regular component that does not use a Context. This makes it much easier to test because all of the values used within the component are passed in explicitly:

```jsx
{/* Easy to test; you can just pass in `highlighted` */}
<Text highlighted={new Set(['1', '2'])} />
```

---

The `src/lib/components/Treebank/Text/Text.js` component imports the context and the `src/lib/components/Text/Text.js` component then returns a component whose values are taken from the context. (This type of component is called a React Context Consumer.):

```jsx
<Sentence highlighted={['1', '2']}>
  {/* Hard to test; you need to create a whole `<Sentence>` component */}
  {/* The `<Sentence>` component may require values or setup unrelated to `<Text>` */}
  <Text />
</Sentence>
```

irina060981 replied:
If to be honest - it is really confusing to see this:

![image](https://user-images.githubusercontent.com/12377640/111617299-a8fda380-882e-11eb-816d-bb9c9bc46bef.png)

and the same inside Treebank
![image](https://user-images.githubusercontent.com/12377640/111617334-b450cf00-882e-11eb-9891-81059d03e286.png)

I don't know React well, but do you really need this duplicated structure?

zfletch replied:
This structure looks confusing, but I think it's the best way to organize code using React Contexts. Consider two hypothetical components: one of them displays a list of highlighted words, the other one shows grammatical information related to the active word. We would use them in an app like this:

```jsx
<Sentence active="4" highlight={["1", "3", "10"]}>
  <HighlightedWords />
  <ActiveWordGrammar />
</Sentence>
```

The `<Sentence>` component passes `active` and `highlight` to the children. But now, let's say we wanted to test the `<HighlightedWords>` component. We would have to write the tests like this:

```jsx
<Sentence active="4" highlight={["1", "3", "10"]}>
  <HighlightedWords />
</Sentence>
```

In order to test `<HighlightedWords>`, we would have to create a whole `<Sentence>` component and everything it requires. But this is a problem. We're adding a bunch of things to our tests unrelated to the functionality that we actually want to test. When testing `<HighlightedWords>`, we shouldn't have to care about `<Sentence>`, and we definitely shouldn't have to care about `active`. 

To address this, we create a separate `<HighlightedWords>` component that does not rely on `<Sentence>`. We can now test it more easily:

```jsx
<HighlightedWords highlight={["1", "3", "10"]} />
```

Doing it this way makes testing easier, but more importantly, it makes the code easier to reason about (making things easier to test often has this side effect). Each component is much simpler and only uses the props given to it.

The only problem is that we have to now create a component that connects the `<Sentence>` and `<HighlightedWords>`. This makes the code look duplicated, but these components are mostly bookkeeping and don't add any functionality of their own.

irina060981 replied:
I am not well-familiar with React and couldn't suggest any other way to organize in such cases.
There are several approaches on this points - should we test the code that is currently used in the application or should we make some adaptations for testing. This is a question what is the priority - test-driven development or first is optimized development and then testing.
I like more independent design in the code and try to avoid global parameters. 
But I am not sure if it is the same situation here. And anyway your code is easy to read - so it is well-organized internally :)

May be it is worth to add such an explanation somewhere in readme for tests?

zfletch commented on src/lib/components/Text/Text.js:
```diff
@@ -41,6 +41,10 @@ const wordToSpan = (word, config, active, toggleActive) => {
     classes.push(styles.active);
   }
 
+  if (highlighted.has(id)) {
```
This is where highlighting is added to highlighted words. For every word in the sentence, the `highlighted` Set is checked for that word's ID. If it is present, the `highlighted` class is added.

The reason that this is a Set instead of an array, is that the `Set.has` function is performed in constant time, whereas the `Array.includes` function takes linear time. In other words, `Set.has` is done in a single computation, while `Array.includes` performs `Array.length` computations.

So, hypothetically, if there were a sentence of 200 words and 100 of them were highlighted, this code would take roughly 200 computations (it runs once for every word in the sentence; 200 X 1 = 200). If the code used an array, then for every word, the `Array.includes` function would take an average of 100 computations. This would result in 200 X 100 = 20,000 computational steps, two orders of magnitude greater than the code using a set.

irina060981 replied:
Thank you for a such detailed exaple here - I also found an article that confirms your calculations about include inside another array
https://dev.to/arnaud/using-array-prototype-includes-vs-set-prototype-has-to-filter-arrays-41fg

Would think about my code

zfletch commented on src/lib/components/Text/Text.module.scss:
```diff
@@ -17,12 +17,16 @@
+}
+
+.highlighted {
+  background-color: #b8f5b9;
```
This is something I'd really like feedback on. I'm not very good with colors. Does this color (`#b8f5b9`) look like a good highlighting color? Is there be a better color to use? The only requirements are that it be different from active (`#f6d918`) and hover (`#5bc8dc`) and that the foreground text be readable.

irina060981 replied:
I left my feedback

irina060981 commented:
@zfletch , thank you for such detailed comments!

**My feedback from user perspective:**
- After reading the issue I expected that highlighting would have the same logic as activating a single word; several words are highlighted in a sentence and they are highlighted in a tree; this way - a user could see a specific subtree. But now only words in sentence are highlighted. And after activating them - there styles are mixed with active words.
- If a user would activate a highlighted word in a treebank then no activate styles would be applied to a highlighted word in a sentence.
This all is rather confusing.

I believe there are several use cases for a word:
**In a sentence**
- !highlighted && !active && !hover - background: transparent
- highlighted && !active && !hover - background: `#b8f5b9`
- highlighted && active && !hover - background: `#f6d918`
- !highlighted && active && !hover - background: `#f6d918`
- highlighted && active && hover - background: `#5bc8dc`
- !highlighted && !active && hover - background: `#5bc8dc`
Also there is a border for active and not hovered

and may be the same for treebank?
I think that there should be a priority - highlighted (1), hover(2), active (3)  for the colors.

I suggest to use monochrome colors for highlighted and active:

Define highlighted colors (I got the analogous color `#f6d918`, i use https://www.color-hex.com/color/a4f618 for such choices):
you also use border for active - I suggest to use monochrome darker for active border

highlighted - `#a4f618` 
highlighted && hover - `#c1f962` (lighter monochrome)
border for active - `#78ba07` (darker monochrome)

active- `#f6d918` 
active&& hover - `#f9e562` (lighter monochrome)
border for active - `#f9e562` (lighter monochrome)

And I suggest to use the same background/border highlighting for sentence and treebank
Hope it would be useful.

irina060981 reviewed (changes requested):
From my point of view you should define background color definition more straightforward for clicking/hovering in text and in treebank before merging.

zfletch commented:
@irina060981 thanks for the excellent code review!

> After reading the issue I expected that highlighting would have the same logic as activating a single word; several words are highlighted in a sentence and they are highlighted in a tree; this way - a user could see a specific subtree.

Great point! I've added equivalent highlighting logic to both the `<Xml>` component and the `<Graph>` component.

> If a user would activate a highlighted word in a treebank then no activate styles would be applied to a highlighted word in a sentence.

Good find: this is a bug. If a word is active, it should show the active style even if it's highlighted. I've updated the code so that active words always show active style (except when hovered over). 

> I use https://www.color-hex.com/color/a4f618 for such choices ... I suggest to use monochrome darker for active border

Thank you for the suggestion and link. I've updated the style to use a darker monochrome. The borders are also a good suggestion. I'm going to leave them out of this PR, since it already includes several changes, but keep the idea in mind for [future styling updates](https://github.com/perseids-tools/treebank-react/issues/31).

The other change I've made is renaming the `highlighted` prop `highlight`.
